[Form Recognizer] Initial design document v2

kristapratico opened the pull request.

mayurid commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+string url/path to the image, or as a file stream. The SDK will determine content-type and send the appropriate header. 
+
+Both prebuilt methods return poller objects which are used to get the result.
+The `begin_extract_receipt` method returns an `ExtractedReceipt` with hardcoded receipt fields.
```
I understand this is something .NET is looking to add to the UX study and make sure the naming resonates with customers trying the API. I am ok changing this accordingly.

annelo-msft replied:
We will definitely ask questions about this in our user study.  Our hypothesis is that Extract will be preferred, given that it is a more specific verb than Analyze.  If we learn this to be incorrect, we can incorporate those learnings.

johanste commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+
+receipt_image = "https://raw.githubusercontent.com/Azure-Samples/cognitive-services-REST-api-samples/master/curl/form-recognizer/contoso-allinone.jpg"
+poller = client.begin_extract_receipt(receipt_image)
+receipt = poller.result()
```
The python client libraries for Azure are using a poller for all long running operations, where long running is defined as something where the client is expected to poll for the eventual result. The pattern is not uncommon in client libraries for services/networking libraries.

Is there something specific that you feel is not idiomatic for Python?

mayurid replied:
A few more reference point from guidelines and existing library.
https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/keyvault/azure-keyvault-keys/samples/backup_restore_operations.py#L53 
https://azure.github.io/azure-sdk/python_design.html#python-lro-prefix

mayurid commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+`CustomFormClient`, which can be imported from the `azure.ai.formrecognizer` namespace. The asynchronous clients 
+can be imported from `azure.ai.formrecognizer.aio`.
+
+`FormRecognizerClient` provides methods for interacting with the prebuilt models (receipt and layout).
```
@ramparab : Just an correction on the comment? Any design issues?

annelo-msft replied:
@ramparab we had discussed this a bit over email, and I'm not sure I understood your answer, so perhaps we can clarify here?  I understand that there are prebuilt and custom models.  It is clear that layout is not a custom model in the sense that a user did not train it by providing their own input data.  If it's not a prebuilt model, how should we understand it?  I'd like for the SDK team to understand this, so we can make sure the documents we put together for our libraries can explain this for customers.

bojunehsu commented:
I like the overall goal of providing a higher level interface for prebuilt models to simplify the developer experience.  However, I believe we may have missed a few common usage scenarios.  I also have a few questions around how the design works with API/model versioning.  I will share more details in a bit.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    value: LabelValue
+
+class LabelValue:
+    text: str
```
A `LabelValue` is technically a `FieldValue` which can have typed fields. Today they are all strings (for custom labeled), but we should leave room for types.

kristapratico replied:
That's good to know. Will sync with team about this so we leave room for types

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+class CustomModel:
+    model_id: str
+    status: str
+    created_date_time: ~datetime.datetime
```
How are other SDKs naming their datetimes? Would it make sense to call this just `created` or `created_on`?

kristapratico replied:
Good point, a quick review of other SDKs shows that `created_on` would be more consistent naming.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+Both methods take as input a blob SAS uri or valid path to the documents to use for training. Each training method 
+will return a poller object which is used to get the training result.
+
+A custom model can be used to analyze forms using the `begin_extract_form` or `begin_extract_labeled_fields` methods.
```
The "with labels" model will also extract tables. Is `begin_extract_labeled_fields` too specific?

kristapratico replied:
Good point. With the recent design changes of merging the result type, the custom form analysis methods will likely collapse into one `begin_extract_form` method. I'll update this design soon.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+
+blob_sas_url = "xxxxx"  # form to analyze uploaded to blob storage
+poller = client.begin_extract_labeled_fields(blob_sas_url, model_id=custom_model.model_id)
+result = poller.result()
```
How do I get the train result later? Should we have a `client.get_training_result(model_id)` method?

kristapratico replied:
Can you speak to the user scenario a little bit here?

Our guidelines say that the poller object should be restart-able to get the result. Is there a scenario of needing to get the train result later (when poller obj is not available)? If so, I'll definitely add this method.

ctstone replied:
There is a lot of model metadata on this result object. To name just one scenario, I might want to generate a report of model accuracies for all of my models.

From reading the C# guidelines, it sounds like I can call `client.begin_extract_labeled_fields(model_id).get_value()` to retrieve the model metadata. This might work, but it's pretty cumbersome for a simple GET operation, and it's not intuitive that `begin_xyz` should be called to get an existing resource.

kristapratico replied:
Thanks for the insight. I will add this method to design doc and we can discuss more at arch board.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+The `begin_extract_layout` method returns the extracted tables in a tabular format such that the user can
+index into a specific row or column and easily integrate with other Python libraries.
+
+If the keyword argument `include_text_details=True` is passed in, the `raw_` attributes will be populated with the
```
Sorry if unclear - I meant that any attribute prefixed with `raw_` would include the re-hydrated readResult referenced by `elements`.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+
+client = FormRecognizerClient(endpoint: str, credential: Union[CognitiveKeyCredential, TokenCredential])
+
+client.begin_extract_receipt(form: Any, **kwargs) -> LROPoller -> List[ExtractedReceipt]
```
Good question - I think it could make sense to include a `page_metadata` field on the receipt that would include this data.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+
+client = FormRecognizerClient(endpoint: str, credential: Union[CognitiveKeyCredential, TokenCredential])
+
+client.begin_extract_receipt(form: Any, **kwargs) -> LROPoller -> List[ExtractedReceipt]
```
Thanks, I meant to narrow down the typing here. Will update to `Union[str, BytesIO]`

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
````diff
@@ -0,0 +1,436 @@
+
+### Receipt Models
+```python
+class ExtractedReceipt:
````
Will update with page_range

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    transaction_time: str
+    fields: ReceiptFields
+
+class ReceiptItem:
```
Thank you, I did miss this. It looks like this is the price per item, so I will add an `item_price`

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    transaction_date: FieldValue
+    transaction_time: FieldValue
+
+class ReceiptItemField:
```
Maybe I'm misunderstanding, but I do see these values being populated on a receipt item object: [example](https://github.com/ctstone/azure-sdk-for-net/blob/form-recognizer-v2.0-GA/sdk/formrecognizer/samples/json/analyze-receipt.json#L172). 

Should we be planning for these values not being populated in the future?

ctstone replied:
The subfields inside the receipt item have these values, but the only properties on the valueObject itself are its field names:

```json
{
    "type": "object",
    "valueObject": {
        "Quantity": { },
        "Name": { },
        "TotalPrice": { }
    }
}
```

kristapratico replied:
Ah I misunderstood your original comment. I will fix this type incorporating your feedback here and below

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    transaction_date: FieldValue
+    transaction_time: FieldValue
+
+class ReceiptItemField:
```
I've since realized this, thanks for pointing it out.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    confidence: float
+    raw_field: List[ExtractedLine]
+
+class FieldValue:
```
Makes sense, will add `text` back.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    value: Union[str, float, int]
+    bounding_box: List[float]
+    confidence: float
+    raw_field: List[ExtractedLine]
```
Will re-think naming on this one.

I'm hoping to return a homogeneous list here to simplify things. My understanding of the readResult response is that  the `elements` field references a Line which is composed of Words. When would I only need to return the Words and not the Line they came from?

ctstone replied:
The `elements` field may reference a Line or a Word (word is more common currently).

Think of a hypothetical receipt line item. The line text might be:
`1  $5.00  Slice of Pizza`

For fields Quantity, Price, and Name, respectively.

The words in that line would be:
`["1", "$5.00", "Slice", "of", "Pizza"]`

So the `Name` field would have an `elements` array of:
```json
[
  "#/readResults/0/lines/0/words/2",
  "#/readResults/0/lines/0/words/3",
  "#/readResults/0/lines/0/words/4",
]
```

If the Name referenced the entire line, it would point to extra/incorrect information related to other fields.

kristapratico replied:
Very clear, thanks for the example. Will update this accordingly

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    column_index: int
+    column_span: int
+    confidence: float
+    is_footer: bool
```
I see these fields come back on the unsupervised result

ctstone replied:
I wasn't sure if this data structure was only for Layout. If it's used by the unsupervised method too, then it makes sense to keep the values.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+
+class TrainResult:
+    documents: List[TrainingDocumentInfo]
+    average_model_accuracy: float
```
Good catch! I'll remove `average_model_accuracy`

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    training_errors: List[ErrorInformation]
+
+class TrainingDocumentInfo:
+    document_name: str
```
I think I was trying to separate the document errors from overall training errors, but will re-think naming to include status/pages.

ctstone replied:
For the record, I prefer the short names, `name`, `errors`, `status`, ~~`pages`~~, `pageCount`

`document.document_name` is a little unwieldy :)

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+
+Both prebuilt methods return poller objects which are used to get the result.
+The `begin_extract_receipt` method returns an `ExtractedReceipt` with hardcoded receipt fields.
+The `begin_extract_layout` method returns the extracted tables in a tabular format such that the user can
```
Currently this is a list of lists. I think there's a few ways we can implement this to make it index-able into a specific cell. Let me look into this a little bit and get back to you - I think the exercise of writing a few samples handling the column/row span scenario might help see what makes sense here.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    total: float
+    transaction_date: str
+    transaction_time: str
+    fields: ReceiptFields
```
These fields will come back in the deserialized response via the gen code. Is this a perf concern? Or are 
you saying to hide these fields to not overwhelm the user?

ctstone replied:
Yes, mainly perf, since the information will be duplicated (or triplicated since this is a class object and the original response is presumably sitting in a dict).

What do you think about using generics here?

johanste replied:
We will not duplicate the values. The only (memory) overhead would be bookkeeping data. We'll measure this and make sure that we don't have issues/understand the overhead. 

What are you referring to with "generics"?

ctstone replied:
Sure, for generics:

`a`, `b`, and `c` are statically typed to `int`, `str`, and `datetime`, respectively. It's still dynamic typing, but the static hints allow a sensible IDE or analysis tool to understand the value type.

```python
from datetime import datetime
from typing import Generic, TypeVar

T = TypeVar('T')


class FieldValue(Generic[T]):
    def __init__(self, value: T):
        self.value = value


if __name__ == '__main__':
    field1 = FieldValue(value=123)
    field2 = FieldValue(value='hello')
    field3 = FieldValue(value=datetime.utcnow())

    a = field1.value
    b = field2.value
    c = field3.value
```

This was written in 3.7, but I think it should work as far back as 3.5 too.

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    quantity: int
+    total_price: float
+
+class ReceiptFields:
```
We can make `ReceiptFields` dict-like in order to have this capability

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    page_number: int
+
+class ExtractedTable: 
+    List[List[TableCell]]
```
Looking into this, will get back to you

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+# Form Recognizer Design Overview
+
+The Form Recognizer client library provides two clients to interact with the service: `FormRecognizerClient` and 
```
I don't think have a strong opinion on this, hoping ux studies might shed some light on discoverability of clients/methods. I like the idea of a `PrebuiltFormClient` for the prebuilt models. If this will grow to many different models/types then it makes sense to split into specific clients, e.g. `ReceiptClient`.

The latest design changes from .NET have the merged `GetAnalyzeResult` outputs so separating custom form clients
might not work here. Before preview we will align client offerings across libraries.

johanste replied:
How is the service planning on communicating which "feature" of the service that customers should use? And what is the projected future usage ratios (if different than future usage ratios)? 

I think this is a good discussion to have. One thing to call out is that, for client libraries, we generally consider usage as "number of users making the call" whereas if often is "number of service calls" when usage is reported from the service side.

ctstone replied:
1. We expect users to start with our documentation and samples: https://docs.microsoft.com/en-us/azure/cognitive-services/form-recognizer/overview
2. We do not anticipate a change in usage ratios in the immediate future.
3. It's an interesting point, and I don't want to skew our analysis. How can we be sure which is the "right" measurement? I'll work with Kusto to see if we can better understand this, but it would help to set some constraints. E.g. what is a "user"? Is it an Azure Resource/ResourceGroup/Subscription? Should we include UserAgent somehow?

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+    source: str, source_prefix_filter: str, include_sub_folders: bool=False
+) -> LROPoller -> CustomModel
+
+client.begin_extract_form(form: Any, model_id: str,) -> LROPoller -> List[ExtractedPage]
```
Just to clarify - are you saying to include the `readResult` per page in addition to the the `raw_fields` (which contain the rehydrated `readResult` from the `elements` ref)?

ctstone replied:
Yes, that's right. `readResult` has important metadata about the page, like dimensions and orientation.

kristapratico replied:
How do you feel about adding these fields as `page_metadata`? Or is there a use case of returning the entire `readResult` on the page?

ctstone replied:
I guess if I went to the trouble of requesting `include_text_details` I would expect to see the full API response.

Scenarios:
1. I am writing an app that compares the full OCR response (all detected words/lines) against the detected keys/values/tables/documents to see if we can improve our annotations.
2. I am building a search engine and I want to capture both structured data (keys/values/tables/documents) and unstructured data (raw words) in my index.
3. I am building a UI that displays keys/values/tables/documents from analyzeResults, but my users need to see the low level words that were detected so they can understand if they have a recognition problem (missing words) or a semantics problem (missing the meaning of the words).

I think it would be a mistake to omit the full `readResult` content from the SDK, and as a user I would be frustrated if there were parts of the Service API that I could not access from the SDK.

johanste replied:
Building client libraries is a constant question of engineering trade-offs. Many of which are hard. We have had many discussions on if it is worth making the API slightly less user friendly for the 99% use case in order to greatly simplify the 1% use case.

We generally follow the path of "make the easy things easy and make the hard things possible". I don't dispute that the scenarios above are possible. But I would like to get a read as to what percentage of applications do it. And what someone would already have had to understand to use the information. And, at that point, how much of a hinderance is the proposed SDK design going to be.

ctstone replied:
I agree with "make the easy things easy and make the hard things possible" but if we omit part of the API service response (and correct me if I am misunderstanding the proposal) then we are making the hard things _not_ possible using our SDK, right? What is the penalty for including a `page_info` (or whatever) property on the model?

Of course the counter argument to "follow the users as we find them" is that we will end up building a nicer horse carriage instead of a car :)

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,436 @@
+class LabeledExtractedPage:
+    fields: List[ExtractedLabel]
+    tables: List[ExtractedTable]
+    page_number: int
```
I can't find what a `fieldRange` is. Do you possibly mean `pageRange` or `fieldName`?

ctstone replied:
Sorry! Yes, I meant `pageRange` here.

adxsdk6 commented:
Can one of the admins verify this patch?

kristapratico commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,496 @@
+) -> LROPoller -> LabeledCustomModel
+
+client.begin_training(
+    source: str, source_prefix_filter: str, include_sub_folders: bool=False, include_keys: bool=True
```
remove include_keys as a param, default to True

bojunehsu commented:
Can't figure out how to comment inline.  Here are feedback on the Prebuilt/Layout portions of the design.

* L5: What is the difference between azure.ai.formrecognizer and azure.ai.formrecognizer.aio namespaces?  I thought we are not providing synchronous clients.
* L11: Not sure if our API supports token authentication yet.
* L7: In the long term, I anticipate the majority of usage to come from customized/adapted models, rather than prebuilt ones.  We also anticipate increasing the number of prebuilt models over the coming years (receipt, business cards, invoice, driver’s license, passport, tax documents, etc.).
* L7: It is unclear how the current design of having begin_extract_{receipt/business_card} will scale.  Do we want to have all the Extracted{Receipt,BusinessCard,Invoice,…} classes be listed under a single namespace?  Or do we want to create sub namespaces such as azure.ai.formrecognizer.receipt?  If so, would it be easier to just have ReceiptClient and BusinessCardClient and have custom forms methods under CustomFormClient (or FormClient)?
* L7: Personally, it feels slightly awkward to group begin_extract_layout with other prebuilt extractions.  Unless we have completely separate clients, I don’t have a better suggestion though.
* L16: For URL input, the expectation is that it is a publicly accessible URL.  From SDK’s perspective, I suspect most developers would expect to be able to pass in a local path.  Not sure what is the best way to handle this expectation gap.  (In the long term, we need to revisit this and allow URLs to Azure Storage with SAS token so the URL does not have to be completely public.  Arguably, we should consider allowing S3/Google Drive/DropBox/… URLs with appropriate authentication tokens.)
* L16: We changed the v2 API design to accept specific content-types (ex. image/png) instead of using generic application/octet-stream or multipart/form-data (v1-preview) with the goal of making this “easier” to use.  In practice, it appears that this is actually harder to use as developers need to figure out the image type from the binary file.  For now, having auto-detect in the SDK is good.  In the future, we should look into passing the binary data using Base64 in the JSON request, where we can optionally specify the content type.
* General: FR Layout and CV Read currently have a large overlap in their response structure.  They share effectively the same underlying backend and differ primarily by the additional table (and soon checkbox) extraction as well as support for smaller fonts.  I assume there is no plan to consider using a common set of data structure classes.  (I don’t recommend sharing at this time.)
* L22: In addition to hydrating elements, include_text_details=True is needed for page_metadata.lines to be populated.
* L38: This current design doesn’t provide an easy way to enumerate all the extracted fields without checking the type.  Do we want to consider ExtractedReceipt.fields.merchant_name?
* L38: Note that for business cards, we will have a full_name field that further contains first_name and last_name.  All 3 fields specifies value, text, bounding_box, confidence, elements.  One approach to represent this is to add dict(string, FieldValue) as a FieldValue.value type, which would defeat the purpose of this high level API.  Another possibility is to simply flatten such nested fields so that we have ExtractedBusinessCard.{full_name,first_name,last_name,name_prefix,name_suffix}.
* L51: A receipt may span multiple pages.  So I assume we need List[PageMetadata] or equivalent.  Naming should be adjusted.
* L53: In the future, we may consider providing confidence and bounding_box/page_number/elements information for each ReceiptItem, pending customer feedback.
* L65: We currently only return ExtractedWord.  How can we introduce ExtractedLine in the future without breaking change?  How can we introduce ExtractedCheckbox in v2.1 without breaking change?  Do we need a base interface instead of union (or is this not a Python concept)?
* L62: I would recommend creating a dedicated BoundingBox class with helper properties since it is a pain for developers to figure out what the 8 numbers in List[float] means.
* L70/L87/L154/L171: Please remove ExtractedLine.language and PageMetadata.language.  They are currently optional and we will remove them in v2.1 because the “language” associated with a line is too contextual and there often is no right answer.
* L71/L79: Using elements[0].line.page_number is one way to capture the page number information (slightly roundabout in my opinion).  When we need to handle checkboxes in v2.1, I assume it will look like elements[0].page_number, since checkbox is planned under /pageResults/*/checkboxes/*/.  If this is .NET, I would still expect to use an interface for elements that directly contains page_number and bounding_box, since this is the primary usage scenario for elements.
* L84/L85: Width/height need to be float since PDF uses inches as unit.
* L86: Unit should be an enum of Pixel and Inch.
* L99: Be careful uploading images to public GitHub repos.  This example image is fine since we created it with all fake data.  We most likely cannot publish any real data due to PII/legal considerations.
* L101: Consider renaming to receipts.
* L122: This would loop over page_range and page_metadata and fail (unless we modify the default behavior of items() somehow).  See comment for L38.
* L130: Note that these are layout tables, not semantic tables which may span across pages.  We will likely need to introduce semantic tables in the future that span across pages.  One way is to add ExtractedTable.{previousTable,nextTable}, which is arguably tricky to use.  Another is to introduce a new semantic_tables property off of each extracted document.  Unfortunately, the current design more or less precludes this as begin_extract_layout returns List[ExtractedLayoutPage].

bojunehsu commented:
* L198: Notes on custom forms.  Currently, we are splitting the clients based on how the model is created, rather than the usage/output of the model.  In the future, we anticipate a customer scenario is to take a prebuilt receipt model and customize it with customer specific data/labels, such as ones from a specific shopping mall.  In such scenario, it still makes sense to return the results via the ExtractedReceipt object, even though the model is strictly speaking custom  Further, building/training a model (from scratch or prebuilt) is often a separate process from using a model for inference/prediction.  Would it make sense to apply something similar to the builder design pattern, where we have different classes for building the model?  This way, when (not if) we have more advanced techniques to train(v1.0)/combine(v2.1)/adapt(v2.x) models, it doesn’t impact the inference codebase, which is generally more widely deployed and more difficult to update.  If desired, I can provide possible options here.
* L202: How would SDK work against on-premise container deployments?  What endpoint would we pass to the client?  What is considered “valid path”?  @Chris Stone
* L205: I am not really a fan of separating begin_extract_forms and begin_extract_labeled_forms, as I don’t really know which one I should call.  Previously, we separated the methods because we believe the output is very different.  But starting in v2.1, we will be including the keyValuePairs/labelValues in the supervised form results as well.  If we want a stable SDK, I would recommend merging the methods and response now.
* L206: If we want an object oriented API, should we have a proper Model object that we can query for details and use to extract form fields?  The model object can be constructed from a modelId, returned from train* methods, or selected from a set of prebuilt models.  Prebuilt models include receipt, business cards, invoices, etc.  Further, there may be multiple models of receipt, corresponding to different model versions.  I can provide options here if desired.
  * I noticed that we do have CustomModel and CustomLabeledModel.  Why can’t we pass this object to begin_extract_forms, or better, move begin_extract_forms to the model classes?
* L209: The current design feels unorganized to me.  It feels like we are exposing the REST APIs directly from CustomFormsClient without introducing object oriented patterns.  From my perspective, each Form Recognizer subscription may contain multiple models.  We can enumerate the models and query model-specific information from the model object.  Internally, we may not want to map to REST API so directly here so the calls are more chunky.  I can propose if desired.
* L220: Is source_prefix_filter optional?  I would expect it to default to “”.
* L230: Similar to receipts, supervised form return list of recognized documents, each containing a set of recognized field values.  The current design only allows us to return a single document.  In addition, supervised form returns OCR/layout information per page, which are not directly tied to each document since we don’t guarantee that a single page maps to exactly one document.  Further, in v2.1, begin_extract_labeled_forms will return keyValuePairs as well in response to customer feedback.  I would recommend following the REST API response JSON more closely or consider Option 3 from the previous mail.
* L254-271/L318-343: As a developer, I am personally confused by this design and the current naming (which I know follows REST API design).  I don’t understand why we have two sets of model results.  I am also concerned that the current design for supervised form will not generalize when we introduce ModelCompose feature in v2.1 (very soon).  Let me follow up with others on the form recognizer team to determine our options.  My apologies that the current design, including REST API, may not work because we didn’t plan far enough.
* L348: It is awkward to have ExtractedForm.pages since a single page may not belong to exactly one form document.  An empty page will not belong to any document, but we still would like to know the basic metadata.  In some cases, a single page may contain multiple documents (ex. receipts).  If this is just a convenience pointer and the information is repeated at a higher level, this is fine.
* L391: I wonder if we should have a Model object with a refresh method.
* L397: I see ModelsSummary as describing the models within the current FR subscription.  Perhaps we can rename this to ClientModels or SubscriptionModels and add a refresh method.  Alternatively, it may make sense for the model count/limit info to be placed directly under CustomFormClient.  It feels like we are currently halfway between autogen APIs based on REST and a truly object oriented API that is designed from scratch.
* L414: I assume we can retrieve model_id from begin_training without having to wait for it to complete.  Since training may take a while, we may not wait for the process to complete and may query the results in a subsequent process.
* L431: Something looks funny here.  Each unsupervised training run should yield a set of clusterIds (form_type_id), not a single one.
* L475: In v2.1, we will have multiple set of fields, one for each document type.  I will work with the FR team to figure out our options.
* L483: Can we call this fields instead of labels?  Supervised does not extract labels and label can also be confused with “data labeling”.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+    item_price: FieldValue
+    total_price: FieldValue
+
+class FieldValue:
```
Coming back to generics. I know it's just a type hint, but we can make `value` explicitly typed. This would replace the Union with `T`.

johanste replied:
Agreed. We'll sort out the type annotations in the actual PR. The high-order bits are 1) generated reference documentation has to look good (both docfx and sphinx generated code), 2) no runtime performance hits (which can happen if you are not careful in how you define you generic types) and 3) we don't accidentally over document the generics (which causes unnecessary breaking changes down the line).

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+    total_price: FieldValue
+
+class FieldValue:
+    value: Union[str, float, int, datetime]
```
Could we use a `datetime.timedelta` for field type `time`?

kristapratico replied:
Why not `datetime.time`? What duration is this representing on a receipt?

ctstone replied:
Yes, `datetime.time` works better. I guess I missed that when I was looking at types (timedelta would have been a duration from midnight)

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+    merchant_address: FieldValue
+    merchant_name: FieldValue
+    merchant_phone_number: FieldValue
+    receipt_type: FieldValue
```
@bojunehsu should `receipt_type` be a `FieldValue` here? Or a string/enum?

kristapratico replied:
receipt_type also includes a `confidence`. Might make sense for this to be its own model, like:
```
class ReceiptType:
    type: str
    confidence: float
```
instead of a FieldValue.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+    transaction_date: FieldValue
+    transaction_time: FieldValue
+    page_range: PageRange
+    page_metadata: PageMetadata
```
What happens if my receipt spans >1 page?
- Which `PageMetadata` would be stored here? The first page of the receipt?
- How would I access the `PageMetadata`'s for the subsequent pages?

kristapratico replied:
Maybe I am not setting up the right sample scenario here, but when I send a receipt that spans more than one page, I get each page of the receipt as a separate item in documentResult: [example](https://github.com/kristapratico/azure-sdk-for-python/blob/fr-json/sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/json-2-page/receipt_2page.json#L5345). This is unlike the result I get from calling a custom supervised method, which will be a single documentResult with page range, e.g. [1, 2].

So I may have represented it wrong if I'm not sending a good multi-page sample to the service, but the idea here would be that we have a `list[ExtractedReceipt]` where each item is a page and each `PageMetadata` would represent the OCR on that page.

ctstone replied:
I think this might be a coincidence b/c each item in `documentResult` has a 1 page span (`[1,1]` and `[2,2]`, respectively). 

What if a single `documentResult` has a `pangeRange` from `[1,2]`? Should `page_metadata` be a list? I don't know if we have an example that reflects this scenario or if it would occur in the wild. @bojunehsu?

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+class FieldValue:
+    value: Union[str, float, int, datetime]
+    text: str
+    bounding_box: List[float]
```
I think .NET is representing bounding_box as a list of 4 x,y points. Should we do the same in Python? cc @annelo-msft

kristapratico replied:
Sounds good, I'll align with the .NET type.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+class ExtractedLayoutPage:
+    tables: List[ExtractedTable]
+    page_metadata: PageMetadata
+    page_number: int
```
Is `page_number` redundant with `page_metadata.page_number`?

kristapratico replied:
Yes, I suppose. But since this class represents a page, I think it makes sense to have this as a top-level attribute.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+    text: str
+    bounding_box: List[float]
+    language: str
+    page_number: int
```
Since `ExtractedWord`'s `line` is a reference to the actual line object, should we make this a reference to the `PageMetadata` object?

kristapratico replied:
This is a good question. The user can always use the page number to find the line on the `PageMetadata` object. If there's a real user need for this, I can add.

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+) -> LROPoller -> CustomModel
+
+# Extract
+client.begin_extract_forms(form: Union[str, BytesIO], model_id: str) -> LROPoller -> List[ExtractedPage]
```
Maybe a personal preference, but I would expect to see model_id as the first parameter since it is a prerequisite.

kristapratico replied:
I don’t think there is any logical ordering or guideline here, but a look at other SDKs shows that the first param passed is generally the one to be “worked on” and anything after that follows.
Storage: https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py#L383

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+# Training ---------------------------------------------------
+class CustomModel:
+    model_id: str
+    status: str
```
Should we use a python enum here (and other status strings)? An alternative might be to expose `bool` properties `succeeded`, `partially_succeeded`, `failed` directly on the object.

kristapratico replied:
Sure, added enums `ModelStatus` and `TrainStatus`

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+    page_count: int
+    errors: List[FormRecognizerError]
+
+class FieldNames:
```
Suggest rename to `FieldName` (not plural) or maybe `FieldInfo`.

kristapratico replied:
Renamed to `FieldInfo`

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+    print(document.page_count)
+    print(document.errors)
+
+print("Form type ID: {}".format(train_result.form_type_id))
```
`form_type_id` should be part of the `extracted_fields[i]` info.

kristapratico replied:
Correct, fixed this in sample

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+result = poller.result()
+
+print("Page range: {}".format(result.page_range))
+for field, value in result.labels.items():
```
I know this is just sample code, but conceptually it might make more sense to think of the dictionary key as the "label" and the value as the "field"

kristapratico replied:
Fixed

ctstone commented on sdk/cognitiveservices/azure-cognitiveservices-formrecognizer/DESIGN_DOC.md:
```diff
@@ -0,0 +1,544 @@
+from azure.ai.formrecognizer import CustomFormClient
+
+client = CustomFormClient(endpoint=endpoint, credential=credential)
+model = client.get_custom_model(model_id="xxxxx")
```
It looks like this returns a `CustomModel` (no labels). Would there be another method to retrieve a `CustomLabeledModel`?

ctstone replied:
Also since we are in a "Custom" Form Client, do we need the qualifier "custom" in `get_custom_model`? Same question applies to delete below.

kristapratico replied:
Yes, there is a `get_custom_labeled_model() -> CustomLabeledModel` method, I added a sample for clarity. I'll reconsider naming with Anne. My only consideration here is making sure the method name is descriptive enough that it can be searched for effectively.

johanste commented:
> L5: What is the difference between azure.ai.formrecognizer and azure.ai.formrecognizer.aio namespaces? I thought we are not providing synchronous clients.

You *must* provide synchronous apis for Python. You *should* also provide asynchronous apis. If you feel like form recognizer falls into the very small number of special cases that would not require a synchronous api, can you please elaborate on why?

 > L7: It is unclear how the current design of having begin_extract_{receipt/business_card} will scale. Do we want to have all the Extracted{Receipt,BusinessCard,Invoice,…} classes be listed under a single namespace? Or do we want to create sub namespaces such as azure.ai.formrecognizer.receipt? If so, would it be easier to just have ReceiptClient and BusinessCardClient and have custom forms methods under CustomFormClient (or FormClient)?

Are you arguing for multiple clients in the same distribution package (same or different namespaces) or for multiple distribution packages? If the expectation is that we'll add many more "pre-trained domains" that all have special/customized return values, and that the user is going to have to specify which domain they want to use, they are going to have to either use a different client, use a different operation on the same client or pass in a parameter to the one and only method on the one and only client that indicates which one to use. 

My recommendation is still to use a single client type and differently named functions on that client. 

> L16: For URL input, the expectation is that it is a publicly accessible URL. From SDK’s perspective, I suspect most developers would expect to be able to pass in a local path. Not sure what is the best way to handle this expectation gap. 

While this may be a concern, we are generally not opening files under the hood for the user (which we'd have to do in order to support local paths). We are using similar patterns in other cases (e.g. blob copy).

> (In the long term, we need to revisit this and allow URLs to Azure Storage with SAS token so the URL does not have to be completely public. Arguably, we should consider allowing S3/Google Drive/DropBox/… URLs with appropriate authentication tokens.)

Wait, that doesn't work today? What are you doing to the URL that I pass to you to make a SAS URI not work?

By the way, you need to figure out how to comment on the PR - this way of responding is not going to scale/work out.

kristapratico commented:
> •  L51: A receipt may span multiple pages. So I assume we need List[PageMetadata] or equivalent. Naming should be adjusted. 

Added a `List[PageMetadata]` on `ExtractedReceipt`.

> •  L84/L85: Width/height need to be float since PDF uses inches as unit. 

Changed to float for `width`/`height`

> •  L230: Similar to receipts, supervised form return list of recognized documents, each containing a set of recognized field values. The current design only allows us to return a single document. In addition, supervised form returns OCR/layout information per page, which are not directly tied to each document since we don’t guarantee that a single page maps to exactly one document. Further, in v2.1, begin_extract_labeled_forms will return keyValuePairs as well in response to customer feedback. I would recommend following the REST API response JSON more closely or consider Option 3 from the previous mail. 

I see what you’re saying. I fixed the return type to be a list and will return all OCR/layout information.

> •  L86: Unit should be an enum of Pixel and Inch. 

Fixed to `LengthUnit` enum.

> •  L220: Is source_prefix_filter optional? I would expect it to default to “”. 

Yes, optional, and defaults to “”.

> •  L122: This would loop over page_range and page_metadata and fail (unless we modify the default behavior of items() somehow). See comment for L38. 

I was originally overriding items(), but I think you're right that it would be more intuitive for users to instead have a `fields` attribute. Added `fields` which is a dictionary of receipt `FieldValue` on the ExtractedReceipt.

> •  L483: Can we call this fields instead of labels? Supervised does not extract labels and label can also be confused with “data labeling”.

Fixed to `fields`
